src/name.h

```diff
@@ -3,42 +3,66 @@
 #define NAME_H
 
 #include <string>
+#include <map>
+#include <vector>
 
-enum nameFlags {
-    nameIsMaleName   = 1 << 0,
-    nameIsFemaleName = 1 << 1,
-    nameIsUnisexName = nameIsMaleName | nameIsFemaleName,
-    nameIsGivenName  = 1 << 2,
-    nameIsFamilyName = 1 << 3,
-    nameIsNickName   = 1 << 4,
-    nameIsTownName   = 1 << 5,
-    nameIsFullName   = 1 << 6,
-    nameIsWorldName  = 1 << 7
+
+enum NameFlags
+{
+    NAME_IS_MALE_NAME,
+    NAME_IS_FEMALE_NAME,
+    NAME_IS_UNISEX_NAME,
+    NAME_IS_GIVEN_NAME,
+    NAME_IS_FAMILY_NAME,
+    NAME_IS_NICK_NAME,
+    NAME_IS_TOWN_NAME,
+    NAME_IS_WORLD_NAME
 };
 
-namespace Name
+class RandomName
 {
-/// Load names from given json file to use for generation
-void load_from_file( const std::string &filename );
 
-/// Return a random name given search flags
-std::string get( nameFlags searchFlags );
+private:
 
-/// Return a random full name given gender
-std::string generate( bool is_male );
+    std::vector<std::string> namesMale;
+    std::vector<std::string> namesFemale;
+    std::vector<std::string> namesUnisex;
+    std::vector<std::string> namesGivenMale;
+    std::vector<std::string> namesGivenFemale;
+    std::vector<std::string> namesFamily;
+    std::vector<std::string> namesNick;
+    std::vector<std::string> namesTown;
+    std::vector<std::string> namesWorld;
 
-/// Clear names used for generation
-void clear();
-}
+    /*
+     * Here will be the instance stored.
+     */
+    static RandomName *instance;
 
-inline nameFlags operator|( nameFlags l, nameFlags r )
-{
-    return static_cast<nameFlags>( static_cast<unsigned>( l ) | static_cast<unsigned>( r ) );
-}
+    /*
+     * Private constructor to prevent instancing.
+     */
+    RandomName();
 
-inline nameFlags operator&( nameFlags l, nameFlags r )
-{
-    return static_cast<nameFlags>( static_cast<unsigned>( l ) & static_cast<unsigned>( r ) );
-}
+    /*
+     * Private constructor to prevent instancing.
+     */
+    RandomName( std::string pathToFile );
+
+    void loadNames( std::string pathToFile );
+
+public:
+
+    /*
+     * Static access method.
+     */
+    static RandomName *getInstance();
+
+    static RandomName *getInstance(std::string pathToFile);
```

Comment:
What prevents me from calling this many times? What do I have to do the return value (do I need to delete it myself)?

Do get a new object each time. What happens if my function uses this multiple times - does the second call invalidate the value returned from the first call?

This does not seem simpler, it just seems more complex.

I agrees that the current interface is a bit undescriptive. But it could be enhanced without adding another indirection.